fenix/app/src/main/java/org/mozilla/fenix/components/appstate/PreferencesMiddleware.kt

```diff
@@ -0,0 +1,47 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.components.appstate
+
+import androidx.datastore.core.DataStore
+import androidx.datastore.preferences.core.Preferences
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+import mozilla.components.lib.state.Middleware
+import mozilla.components.lib.state.MiddlewareContext
+import org.mozilla.fenix.datastore.updatedDebugDrawerEnabled
+
+/**
+ * Middleware that handles [AppAction] related to top-level user preferences and settings.
+ *
+ * @param debugSettingsDataStore [DataStore] for accessing debugging settings.
+ * @param scope [CoroutineScope] used for writing to data stores.
+ */
+class PreferencesMiddleware(
+    private val debugSettingsDataStore: DataStore<Preferences>,
+    private val scope: CoroutineScope = CoroutineScope(Dispatchers.IO),
+) : Middleware<AppState, AppAction> {
+
+    override fun invoke(
+        context: MiddlewareContext<AppState, AppAction>,
+        next: (AppAction) -> Unit,
+        action: AppAction,
+    ) {
+        handleAction(action)
```

Comment:
Since we can probably assume that the actions here are all disk writes, it might help convey some semantic value to do `scope.launch { handleAction(action) }` and make `handleAction` suspending. It would send a message to future readers that `handleAction` doesn't block `next`.